camel/data_collector/base.py

```diff
@@ -0,0 +1,169 @@
+
+        Args:
+            message (Union[BaseMessage, ChatAgentResponse]):
+                The message to record.
+            role (OpenAIBackendRole): The role of the message.
+            role_name (Optional[str], optional):
+                The name of the role. Defaults to None.
+                Used when message if from user.
+        """
+
+        name = role_name or message.role_name
+
+        self.history.append(
+            CollectorData(
+                id=uuid.uuid1(), name=name, role=role, message=message
+            )
+        )
+        return self
+
+    def _inject(self, agent: ChatAgent) -> Self:
+        r"""Inject an agent.
+
+        Args:
+            agent (ChatAgent): The agent to inject.
+        """
+        name = agent.role_name
+        if not name:
+            name = f"{agent.__class__.__name__}_{len(self.agents)}"
+        if name in [n for n, _ in self.agents]:
+            raise ValueError(f"Name {name} already exists")
+
+        self.agents.append((name, agent))
+
+        ori_update_memory = agent.update_memory
+
+        def update_memory(
+            message: BaseMessage, role: OpenAIBackendRole
+        ) -> None:
+            if self._recording:
+                self.history.append(
+                    CollectorData(
+                        id=uuid.uuid1(), name=name, role=role, message=message
+                    )
+                )
+            return ori_update_memory(message, role)
+
+        agent.update_memory = update_memory  # type: ignore[method-assign]
+
+        return self
+
+    def inject(
+        self,
+        agent: Union[List[ChatAgent], ChatAgent],
+    ) -> Self:
+        r"""Inject agents.
+
+        Args:
+            agent (Union[List[ChatAgent], ChatAgent]):
+                The agent(s) to inject.
+        """
+        if not isinstance(agent, list):
+            agent = [agent]
+        for a in agent:
+            self._inject(a)
+        return self
+
+    def start(self) -> Self:
+        r"""Start recording."""
+        self._recording = True
+        return self
+
+    def stop(self) -> Self:
+        r"""Stop recording."""
+        self._recording = False
+        return self
+
+    @property
+    def recording(self) -> bool:
+        r"""Whether the collector is recording."""
+        return self._recording
+
+    def reset(self, reset_agents: bool = True):
+        r"""Reset the collector.
+
+        Args:
+            reset_agents (bool, optional):
+                Whether to reset the agents. Defaults to True.
+        """
+        self.history = []
+        if reset_agents:
+            for _, agent in self.agents:
+                agent.reset()
+
+    @abstractmethod
+    def convert(self) -> Any:
+        r"""Convert the collected data."""
+        pass
+
+    @abstractmethod
+    def llm_convert(self, converter: Any, prompt: Optional[str] = None) -> Any:
+        r"""Convert the collected data."""
+        pass
+
+    def save(self, path: str):
```

Comment:
I think some sort of built in saving mechanism can be good for very large datasets, though in those cases it probably should be more like a configured save location where results are streamed to rather than just at the end, so a set save location function or something can be better (or maybe we want to use an existing data streaming library, and we could have function which simply gives a stream of results like a torch function or something). So in the meantime maybe we should rename this or delete this